File: go/medium/91_decode_ways.go
// https://leetcode.com/problems/decode-ways/
package medium

import "strconv"

func NumDecodings(s string) int {
	return recNumDecodings(s, 0, make(map[int]int))
}

func recNumDecodings(s string, idx int, memo map[int]int) int {
	if idx == len(s) {
		return 1
	}
	if s[idx] == '0' {
		return 0
	}
	if idx == len(s)-1 {
		return 1
	}

	if _, ok := memo[idx]; ok {
		return memo[idx]
	}

	count := recNumDecodings(s, idx+1, memo)
	c2, _ := strconv.Atoi(s[idx : idx+2])
	if c2 <= 26 {
		count += recNumDecodings(s, idx+2, memo)
	}
	memo[idx] = count

	return count
}

File: java/medium/91_decode_ways.java
// https://leetcode.com/problems/decode-ways/

package java.medium;

class DecodeWays {
    public int numDecodings(String s) {
        int n = s.length();
        int[] dp = new int[n+1];

        // base case
        dp[0] = 1;
        dp[1] = s.charAt(0) == '0' ? 0 : 1;

        for (int i = 2; i < n+1; i++) {
            if (s.charAt(i-1) != '0') {
                dp[i] += dp[i-1];
            }
            if (s.charAt(i-2) == '1' || s.charAt(i-2) == '2' &&  s.charAt(i-1) < '7') {
                dp[i] += dp[i-2];
            }
        }

        return dp[n];
    }
}
